File: add_map_plugin.rb
#!/usr/bin/ruby

require 'nokogiri'

robot_init_x = 0.0
robot_init_y = 0.0

world = ENV['WORLD_ID'] || "aws_robomaker_bookstore_world"

if world=="aws_robomaker_bookstore_world"
  world_file = String.new("simulation_ws/src/deps/aws-robomaker-bookstore-world/worlds/bookstore.world")
elsif world=="aws_robomaker_small_house_world"
  world_file = String.new("simulation_ws/src/deps/aws-robomaker-small-house-world/worlds/small_house.world")
  robot_init_x = 3.0
  robot_init_y = 3.0
else
  puts "Please set a VALID world name, eg: export WORLD_ID=aws_robomaker_small_house_world"
  exit
end

doc = File.open(world_file) { |f| Nokogiri::XML(f) }
_world_ptr = doc.at('world')

plugin_doc = Nokogiri::XML("<plugin name='gazebo_occupancy_map' filename='libgazebo_2Dmap_plugin.so'>
    <map_resolution>0.05</map_resolution> <!-- in meters, optional, default 0.1 -->
    <map_height>0.3</map_height>         <!-- in meters, optional, default 0.3 -->
    <map_size_x>25</map_size_x>          <!-- in meters, optional, default 25 -->
    <map_size_y>25</map_size_y>          <!-- in meters, optional, default 25 -->
    <init_robot_x>#{robot_init_x}</init_robot_x>          <!-- in meters, optional, default 0. -->
    <init_robot_y>#{robot_init_y}</init_robot_y>          <!-- in meters, optional, default 0. -->
</plugin>
")

_world_ptr.add_child(plugin_doc.at('plugin'))

if world=="aws_robomaker_bookstore_world"
  dump_file = String.new("simulation_ws/src/deps/aws-robomaker-bookstore-world/worlds/map_plugin.world") 
elsif world=="aws_robomaker_small_house_world"
  dump_file = String.new("simulation_ws/src/deps/aws-robomaker-small-house-world/worlds/map_plugin.world") 
end

File.write(dump_file, doc.to_xml)
puts "plugin added successfully to " + world_file
puts "new world file saved at " + dump_file + "\n"

File: add_map_plugin.py
import sys
import os
import subprocess

def get_args():
	# default arguments
	if len(sys.argv) == 2:
		default_args = {'bookstore':["./map_config/bookstore.rb", "aws-robomaker-bookstore-world"],\
				'small_house':["./map_config/small_house.rb", "aws-robomaker-small-house-world"],\
				'small_warehouse':["./map_config/small_warehouse.rb","aws-robomaker-small-warehouse-world"],\
				'no_roof_small_warehouse':["./map_config/no_roof_small_warehouse.rb","aws-robomaker-small-warehouse-world"]
				}
		try:
			world_name = sys.argv[1]
			config_file = default_args[world_name][0]
			world_file = "simulation_ws/src/deps/{0}/worlds/{1}.world".format(default_args[world_name][1], world_name)
			output_file = "simulation_ws/src/cloudwatch_simulation/worlds/map_plugin.world"
		except KeyError:
			raise ValueError("Invalid argument")

	#custom world file, config file, output file
	elif len(sys.argv) == 4:
		config_file, world_file, output_file = sys.argv[1:]
	
	else:
		raise ValueError("Expects 1 or 3 number of arguments")

	args = list(map(lambda x: os.path.abspath(x), [config_file, world_file, output_file]))
	
	return args

def main():
	args = get_args()
	cmd = './scripts/genmap.sh '+' '.join(args)
	try:
		out = subprocess.check_output(cmd, shell=True)
	except subprocess.CalledProcessError:
		sys.exit(2)

if __name__=="__main__":
	main()
